app/src/main/java/org/simple/clinic/sync/indicator/SyncIndicatorViewController.kt

```diff
@@ -1,21 +1,53 @@
 package org.simple.clinic.sync.indicator
 
+import com.f2prateek.rx.preferences2.Preference
 import io.reactivex.Observable
 import io.reactivex.ObservableSource
 import io.reactivex.ObservableTransformer
+import io.reactivex.rxkotlin.Observables
+import io.reactivex.rxkotlin.ofType
 import org.simple.clinic.ReportAnalyticsEvents
+import org.simple.clinic.sync.SyncProgress
+import org.simple.clinic.sync.SyncProgress.FAILURE
+import org.simple.clinic.sync.SyncProgress.SUCCESS
+import org.simple.clinic.sync.SyncProgress.SYNCING
+import org.simple.clinic.sync.indicator.SyncIndicatorState.SyncPending
+import org.simple.clinic.sync.indicator.SyncIndicatorState.Synced
+import org.simple.clinic.sync.indicator.SyncIndicatorState.Syncing
+import org.simple.clinic.util.Optional
+import org.simple.clinic.util.filterAndUnwrapJust
 import org.simple.clinic.widgets.UiEvent
+import org.threeten.bp.Instant
 import javax.inject.Inject
+import javax.inject.Named
 
 typealias Ui = SyncIndicatorView
 typealias UiChange = (Ui) -> Unit
 
 class SyncIndicatorViewController @Inject constructor(
+    @Named("last_frequent_sync_succeeded_timestamp") private val lastSyncSucceededTimestamp: Preference<Optional<Instant>>,
+    @Named("last_frequent_sync_result") private val lastSyncGroupResult: Preference<Optional<SyncProgress>>
 ) : ObservableTransformer<UiEvent, UiChange> {
 
   override fun apply(events: Observable<UiEvent>): ObservableSource<UiChange> {
     val replayedEvents = events.compose(ReportAnalyticsEvents()).replay(1).refCount()
 
-    return Observable.never()
+    return updateIndicatorView(replayedEvents)
+  }
+
+  private fun updateIndicatorView(events: Observable<UiEvent>): Observable<UiChange> {
+    val syncResultStream = lastSyncGroupResult.asObservable().distinctUntilChanged()
+
+    return Observables.combineLatest(
+        events.ofType<SyncIndicatorViewCreated>(),
+        syncResultStream.filterAndUnwrapJust()) { _, result ->
```

Comment:
Beware of using a `filter` before `combineLatest`. You'll receive potentially stale values from `combineLatest` if the streams are filtered. 

Tell me if I'm wrong, but it looks like you want to filter after combining the streams here. What happens if the sync result is cleared? Here's my attempted visualization of the streams:

```
viewCreated: SyncIndicatorViewCreated
syncResultStream: None
combineLatest: --

viewCreated: SyncIndicatorViewCreated
syncResultStream: Just(SYNCING)
combineLatest: Pair(SyncIndicatorViewCreated, SYNCING)

viewCreated: SyncIndicatorViewCreated
syncResultStream: Just(SUCCESS)
combineLatest: Pair(SyncIndicatorViewCreated, SUCCESS)

viewCreated: SyncIndicatorViewCreated
syncResultStream: None
combineLatest: Pair(SyncIndicatorViewCreated, SUCCESS)
```

Was I able to explain this?

Reply:
This makes sense. I will do the necessary changes.